tests/models/test_truncated_bptt.py

```diff
@@ -74,104 +60,113 @@ def training_step(self, batch, batch_idx, hiddens):
         opt.zero_grad()
         self.manual_backward(loss)
         opt.step()
-        assert loss.grad_fn is not None
         return {"loss": loss, "hiddens": hiddens}
 
 
-@pytest.mark.parametrize("model_class", (LinearModel, ManualLinearModel))
-@pytest.mark.parametrize("n_hidden_states", (1, 2))
-def test_tbptt_cpu_model_manual(tmpdir, n_hidden_states, model_class):
-    """Test truncated back propagation through time works with automatic and manual optimization."""
+@pytest.mark.parametrize("model_class", (LSTMModel, ManualLSTMModel))
+def test_persistent_hidden_state_transfer(tmpdir, model_class):
+    """Test that the hidden state reference gets passed through from one training_step to the next and remains
+    unmodified apart from detached grad_fn."""
 
-    sequence_size = 30
-    batch_size = 30
-
-    x_seq = torch.rand(batch_size, sequence_size, 1)
-    y_seq_list = torch.rand(batch_size, sequence_size, 1).tolist()
-
-    class MockSeq2SeqDataset(Dataset):
-        def __getitem__(self, i):
-            return x_seq, y_seq_list
+    class TBPTTModel(model_class):
+        def __init__(self, *args, **kwargs):
+            super().__init__(*args, **kwargs)
+            self.test_hidden = None
 
-        def __len__(self):
-            return 1
+        def training_step(self, batch, batch_idx, hiddens):
+            split_idx = self.trainer.fit_loop.split_idx
+            # the hidden state may only be None for the first split_idx
+            assert split_idx == 0 and hiddens is None or split_idx != 0 and hiddens is not None
+            # test_hiddens is None when hiddens is None
+            assert not ((hiddens is None) ^ (self.test_hidden is None))
```

Comment:
Could use XOR here too

and the parentheses aren't necessary

```suggestion
            assert not (split_idx == 0) ^ (hiddens is None)
            # test_hiddens is None when hiddens is None
            assert not (hiddens is None) ^ (self.test_hidden is None)
```

Reply:
yes, added another xnor. 
I think parenthesis are ok here considering probably many people don't know `^` and precedence :)